cpp/src/lists/drop_list_duplicates.cu

```diff
@@ -22,74 +22,129 @@
 #include <cudf/detail/nvtx/ranges.hpp>
 #include <cudf/lists/detail/sorting.hpp>
 #include <cudf/lists/drop_list_duplicates.hpp>
-#include <cudf/table/row_operators.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
 
 #include <rmm/cuda_stream_view.hpp>
 #include <rmm/exec_policy.hpp>
 
 #include <thrust/binary_search.h>
+#include <thrust/logical.h>
 #include <thrust/transform.h>
 
 namespace cudf {
 namespace lists {
 namespace detail {
 namespace {
 using offset_type = lists_column_view::offset_type;
+
+template <typename Type>
+struct has_negative_nans {
+  column_device_view const d_entries;
+  bool const has_nulls;
+
+  __device__ Type operator()(size_type idx) const noexcept
+  {
+    if (has_nulls && d_entries.is_null(idx)) { return false; }
+
+    auto const val = d_entries.element<Type>(idx);
+    return std::isnan(val) && std::signbit(val);  // std::signbit(x) == true if x is negative
+  }
+};
+
 /**
- * @brief Copy list entries and entry list offsets ignoring duplicates
- *
- * Given an array of all entries flattened from a list column and an array that maps each entry to
- * the offset of the list containing that entry, those entries and list offsets are copied into
- * new arrays such that the duplicated entries within each list will be ignored.
- *
- * @param all_lists_entries    The input array containing all list entries
- * @param entries_list_offsets A map from list entries to their corresponding list offsets
- * @param nulls_equal          Flag to specify whether null entries should be considered equal
- * @param stream               CUDA stream used for device memory operations and kernel launches
- * @param mr                   Device resource used to allocate memory
- *
- * @return A pair of columns, the first one contains unique list entries and the second one
- * contains their corresponding list offsets
+ * @brief A structure to be used along with type_dispatcher to check if a
+ * `column_view` has any negative NaN entry
  */
-template <bool has_nulls>
-std::vector<std::unique_ptr<column>> get_unique_entries_and_list_offsets(
-  column_view const& all_lists_entries,
-  column_view const& entries_list_offsets,
-  null_equality nulls_equal,
-  rmm::cuda_stream_view stream,
-  rmm::mr::device_memory_resource* mr)
+struct has_negative_nans_fn {
+  template <typename Type, std::enable_if_t<cuda::std::is_floating_point_v<Type>>* = nullptr>
+  bool operator()(column_view const& lists_entries, rmm::cuda_stream_view stream) const noexcept
+  {
+    auto const d_entries = column_device_view::create(lists_entries, stream);
+    return thrust::any_of(rmm::exec_policy(stream),
+                          thrust::make_counting_iterator(0),
+                          thrust::make_counting_iterator(lists_entries.size()),
+                          detail::has_negative_nans<Type>{*d_entries, lists_entries.has_nulls()});
```

Comment:
The Thrust logical reductions have known performance problems: https://github.com/NVIDIA/thrust/issues/1016

Use a regular reduction like `count_if` instead.